clippy_lints/src/static_items_large_align.rs

```diff
@@ -0,0 +1,92 @@
+use clippy_utils::diagnostics::span_lint_and_note;
+use rustc_hir::{Item, ItemKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::{self, AdtDef, Ty, TyCtxt};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_target::abi::Align;
+use rustc_typeck::hir_ty_to_ty;
+
+pub struct StaticItemsLargeAlign {
+    pub page_size: u64,
+}
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Check for usage of static items which have type alignment larger than page size.
+    ///
+    /// ### Why is this bad?
+    /// Due to some known unsound issues, the type alignment may not be fulfilled.
+    /// For more information, see:
+    /// <https://github.com/rust-lang/rust/issues/70022> and
+    /// <https://github.com/rust-lang/rust/issues/70143>.
+    ///
+    /// ### Example
+    /// ```rust
+    /// #[repr(align(0x100000))]
+    /// struct Aligned(u8);
+    ///
+    /// static X: Aligned = Aligned(0);         // Bad
+    ///
+    /// fn main() {
+    ///     let x = Aligned(0);                 // Good
+    ///     println!("{:#x}", &x as *const _ as usize);
+    ///     println!("{:#x}", &X as *const _ as usize);
+    ///     let b = Box::new(Aligned(0));       // Good
+    ///     println!("{:#x}", Box::into_raw(b) as usize);
+    /// }
+    #[clippy::version = "1.61.0"]
+    pub STATIC_ITEMS_LARGE_ALIGN,
+    pedantic,
+    "static items with large type alignment, which may trigger unsound problems"
+}
+impl_lint_pass!(StaticItemsLargeAlign => [STATIC_ITEMS_LARGE_ALIGN]);
+
+impl LateLintPass<'_> for StaticItemsLargeAlign {
+    fn check_item(&mut self, cx: &LateContext<'_>, item: &Item<'_>) {
+        if_chain! {
+            if let ItemKind::Static(hir_ty, _, _) = item.kind;
+            let ty = hir_ty_to_ty(cx.tcx, hir_ty);
+            if let Some(adt_ref) = self.check_ty_alignment(cx.tcx, ty);
```

Comment:
I did not think about references at all... that's a really good catch. I'm wondering if we'll miss anything with such an approach. I guess we could use the const validator to look at the actual allocations, but not sure we'll have other problems due to that.

Either way, your function looks good and we can add more types to recurse on later.

One thing your PR made me realize is that promoteds suffer from the same issue, but those can even be generic... not sure how to best handle them. Maybe with a mir visitor looking at the types of all promoteds. Doesn't have to be in this PR tho

Reply:
> I guess we could use the const validator to look at the actual allocations

Could you provide more context on const validator and allocations?

> One thing your PR made me realize is that promoteds suffer from the same issue, but those can even be generic...

Your mention of generic reminds me to think more about it when collecting the spans... As for  promoteds, I guess I will take a peek at const eval.

Reply:
The actual memory as we will pass it to llvm can be fetched with https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.eval_static_initializer

This gives you access to an Allocation which has an `align` field. It also gives you a list of relocations, which are ids of other allocations referenced by this static (essentially pointers to other memory). You can then use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.get_global_alloc to fetch the allocations themselves and recurse. This will not give us any information about the spans or similar, but it will make sure we don't miss anything.

We could as a next step employ https://doc.rust-lang.org/nightly/nightly-rustc/rustc_const_eval/interpret/visitor/index.html but that is a bit more involved